src/headers/lnurl_auth_jwt.rs

```diff
@@ -0,0 +1,291 @@
+use crate::headers::get_headermap;
+use crate::headers::VssHeaderProvider;
+use crate::headers::VssHeaderProviderError;
+use async_trait::async_trait;
+use base64::engine::general_purpose::URL_SAFE_NO_PAD;
+use base64::Engine;
+use bitcoin::bip32::{ChildNumber, DerivationPath, ExtendedPrivKey};
+use bitcoin::hashes::hex::FromHex;
+use bitcoin::hashes::sha256;
+use bitcoin::hashes::{Hash, HashEngine, Hmac, HmacEngine};
+use bitcoin::secp256k1::{All, Message, Secp256k1};
+use bitcoin::Network;
+use bitcoin::PrivateKey;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::time::SystemTime;
+use url::Url;
+
+// Derivation index of the parent extended private key as defined by LUD-05.
+const PARENT_DERIVATION_INDEX: u32 = 138;
+// Derivation index of the hashing private key as defined by LUD-05.
+const HASHING_DERIVATION_INDEX: u32 = 0;
+// The JWT token will be refreshed by the given amount before its expiry.
+const EXPIRY_BUFFER_SECS: u64 = 60;
+// The key of the LNURL k1 query parameter.
+const K1_QUERY_PARAM: &str = "k1";
+// The key of the LNURL sig query parameter.
+const SIG_QUERY_PARAM: &str = "sig";
+// The key of the LNURL key query parameter.
+const KEY_QUERY_PARAM: &str = "key";
+// The authorization header name.
+const AUTHORIZATION: &str = "authorization";
+
+#[derive(Debug, Clone)]
+struct JwtToken {
+	token_str: String,
+	expiry: Option<u64>,
+}
+
+impl JwtToken {
+	fn is_expired(&self) -> bool {
+		self.expiry
+			.map(|expiry| {
+				SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs() + EXPIRY_BUFFER_SECS
+					> expiry
+			})
+			.unwrap_or(false)
+	}
+}
+
+/// Provides a JWT token based on LNURL Auth.
+/// The LNURL and JWT token are exchanged over a Websocket connection.
+pub struct LnurlAuthJwt {
+	engine: Secp256k1<All>,
+	parent_key: ExtendedPrivKey,
+	url: String,
+	default_headers: HashMap<String, String>,
+	client: reqwest::Client,
+	jwt_token: RwLock<Option<JwtToken>>,
+}
+
+impl LnurlAuthJwt {
+	/// Creates a new JWT provider based on LNURL Auth.
+	///
+	/// The LNURL Auth keys are derived based on the wallet seed according to LUD-05.
```

Comment:
Yes, I think this is useful for portability following the LNURL Auth spec. I believe we can have access similar to here: https://github.com/lightningdevkit/ldk-node/blob/640a1fdb7833ad9c74ede0926f990d85ac1b3bca/src/builder.rs#L342 ?

We discussed this a little bit on Slack (https://cash.slack.com/archives/C03Q9S7K99R/p1707986857336199?thread_ts=1706544261.279069&cid=C03Q9S7K99R) whether to go with [LUD-05](https://github.com/lnurl/luds/blob/luds/05.md) or [LUD-13](https://github.com/lnurl/luds/blob/luds/13.md) (based on a node signature), but LUD-05 seemed more straight-forward.

Reply:
Would we provide a helper for the parent key derivation to avoid user error? If so, the exposure might be similar? Overall, I felt like giving the wallet seed directly is less error prone and should be safe if the repo is well-reviewed.

Reply:
Is the deviation you mentioned the return of a JWT token from the LNURL Auth server?

The only reason to use the specification would be to get to a point where Lightning wallets can be portable between implementations with just a wallet seed, and have a well-defined derivation from the wallet seed to the linking key to access the protected data on the VSS server side. That might be out of scope right now though, and things will work either way for a single implementation. It just might be difficult to migrate existing VSS data to a different linking key, other things like using JWT are easier to change I think.

Reply:
Yes, added a comment there.

Looking at this again, I'm also down to just take in the parent key as an argument instead, and the user can decide how to derive it, following the spec with `m/138'` from https://github.com/lnurl/luds/blob/luds/05.md or something else.